feat(#1716): Implement receiveBody()

jbonofre opened the pull request.

tabish121 commented on activemq-client/src/main/java/org/apache/activemq/ActiveMQConsumer.java:
```diff
@@ -100,17 +100,44 @@ public void close() {
+        try {
+            Object body = message.getBody(c);
+            return c.cast(body);
+        } catch (JMSException e) {
```
This would violate the JMS specification in that it does not return the message to the head of the prefetch queue to be read again on the next receive call for the AUTO and DUPS_OK ack modes.  Same for the other methods below.

>  The result of this method throwing a MessageFormatRuntimeException depends on the session mode:
> 
>     AUTO_ACKNOWLEDGE or DUPS_OK_ACKNOWLEDGE: The JMS provider will behave as if the unsuccessful call to receiveBody had not occurred. The message will be delivered again before any subsequent messages.
> 
>     This is not considered to be redelivery and does not cause the JMSRedelivered message header field to be set or the JMSXDeliveryCount message property to be incremented.
>

jbonofre replied:
@tabish121 I started the implementation of the body receive failure logic.

jbonofre commented:
@jeanouii thanks for the review. I have to resume my work on this PR. I agree that the approach you are proposing makes more sense (else the ack will be send too early).

I will update accordingly.

@tabish121 @jeanouii thanks for your inputs ! I will resume my work on this one.

cshannon reviewed (changes requested):
I only glanced at this and haven't reviewed the rest of the changes, but one thing that immedaitely stood out that needs to be changed is this completely breaks exception handling. JMSConsumer needs to be converted to a runtime exception but ActiveMQConsumer still implements MessageConsumer. This is a problem because anyone who is using that API may be using a try/catch and looking for JMSException that will now no longer be thrown so it's a major breaking change if client applications are catching checked exceptions that will never be thrown so we can't do that.

I think you need to do something like the approach Artemis did, where they wrap the MessageConsumer instead. See https://github.com/apache/artemis/blob/2e390b4fa0842ed8d3557a158c9ca4d204687cf3/artemis-jms-client/src/main/java/org/apache/activemq/artemis/jms/client/ActiveMQJMSConsumer.java

jbonofre committed:
feat(#1716): Implement receiveBody()

Implement JMS 2.0 / Jakarta Messaging 3.1 receiveBody() methods on
ActiveMQMessageConsumer and wire them through ActiveMQConsumer (JMSConsumer).

The doReceiveBody() helper handles ack-mode-dependent failure behavior
per section 8.6 of the Jakarta Messaging 3.1 specification:
- AUTO_ACK / DUPS_OK: re-enqueue without acknowledgement
- CLIENT_ACK / TRANSACTED: treat as delivered (credit window expansion)

Plain Message and StreamMessage types are always rejected with
MessageFormatException, per spec (even though Message.isBodyAssignableTo
returns true for any type).

ActiveMQObjectMessage.isBodyAssignableTo now returns false on
deserialization failure instead of propagating the exception.

tabish121 commented on activemq-client/src/main/java/org/apache/activemq/ActiveMQMessageConsumer.java:
```diff
@@ -713,6 +714,144 @@ public Message receiveNoWait() throws JMSException {
+     */
+    private <T> T doReceiveBody(MessageDispatch md, Class<T> c) throws JMSException {
+        ActiveMQMessage message = createActiveMQMessage(md);
+        if (!message.isBodyAssignableTo(c)) {
```
You should implement isBodyAssignableTo in each message type to follow the guidance in the API docs.

```
boolean isBodyAssignableTo([Class](https://docs.oracle.com/javase/8/docs/api/java/lang/Class.html?is-external=true) c)
                    throws [JMSException](https://jakarta.ee/specifications/messaging/2.0/apidocs/javax/jms/jmsexception)

Returns whether the message body is capable of being assigned to the specified type. If this method returns true then a subsequent call to the method getBody on the same message with the same type argument would not throw a MessageFormatException.

If the message is a StreamMessage then false is always returned. If the message is a ObjectMessage and object deserialization fails then false is returned. If the message has no body then any type may be specified and true is returned.

Parameters:
    c - The specified type
    If the message is a TextMessage then this method will only return true if this parameter is set to String.class or another type to which a String is assignable.
    If the message is a ObjectMessage then this method will only return true if this parameter is set to java.io.Serializable.class or another class to which the body is assignable.
    If the message is a MapMessage then this method will only return true if this parameter is set to java.util.Map.class (or java.lang.Object.class).
    If the message is a BytesMessage then this this method will only return true if this parameter is set to byte[].class (or java.lang.Object.class).
    If the message is a TextMessage, ObjectMessage, MapMessage or BytesMessage and the message has no body, then the above does not apply and this method will return true irrespective of the value of this parameter.
    If the message is a Message (but not one of its subtypes) then this method will return true irrespective of the value of this parameter.
```

tabish121 replied:
You are correct about the base Message type needing to be accounted for. It is the only case where you have an explicitly defined Message that has "no body" whereas the others might arrive without a body but a null body is itself a body in those cases so the method can return null.

jbonofre replied:
Confirmed — the fix that got lost in an earlier force-push is back in the current commit. `doReceiveBody()` now checks the message type explicitly before trusting `isBodyAssignableTo()`:

```java
boolean bodyNotAssignable = message.getClass() == ActiveMQMessage.class   // plain Message: always reject
        || message instanceof ActiveMQStreamMessage                        // StreamMessage: always reject
        || !message.isBodyAssignableTo(c);
```

This resolves the conflict you pointed out where `Message#isBodyAssignableTo()` must return `true` for any type while `receiveBody()` must reject `Message`/`StreamMessage`. Covered by `testReceiveBodyPlainMessageThrows` and `testReceiveBodyStreamMessageThrows` (23 tests total, all passing).

jbonofre commented:
Reopen for new round.

cshannon commented:
i am going to move this to 6.4.0

jbonofre commented:
@cshannon I didn't define target project on this one as it's best effort. I propose to move forward for now and define the target project (hopefully 6.4.0) when we have a consensus on the implementation.

jbonofre commented on activemq-client/src/main/java/org/apache/activemq/ActiveMQMessageConsumer.java:
```diff
@@ -713,6 +714,144 @@ public Message receiveNoWait() throws JMSException {
+        ActiveMQMessage message = createActiveMQMessage(md);
+        if (!message.isBodyAssignableTo(c)) {
+            // spec: message is not acknowledged and left for redelivery
+            unconsumedMessages.enqueueFirst(md);
```
You were right — a force-push had reverted this. It is restored in the current commit. On a body-type mismatch `doReceiveBody()` now branches on the ack mode:

```java
if (session.isAutoAcknowledge() || session.isDupsOkAcknowledge()) {
    // AUTO / DUPS_OK: put back at the head of the prefetch, no ack
    unconsumedMessages.enqueueFirst(md);
} else {
    // CLIENT / TRANSACTED / INDIVIDUAL: treated as delivered, NOT re-enqueued.
    // beforeMessageIsConsumed + afterMessageIsConsumed send a DELIVERED_ACK,
    // which expands the credit window so the consumer does not stall.
    beforeMessageIsConsumed(md);
    afterMessageIsConsumed(md, false);
}
throw new MessageFormatException(...);
```

So transacted, client and individual-ack modes no longer re-add to the head — the application recovers via `session.recover()` / `session.rollback()`. Both the credit-window-not-exhausted concern and the `Message`/`StreamMessage` type check are now covered by tests (`testReceiveBodyClientAckCreditWindowNotExhausted`, `testReceiveBodyTransactedCreditWindowNotExhausted`, `testReceiveBodyClientAckRecoverRedelivers`, `testReceiveBodyTransactedRollbackRedelivers`, plus the plain-Message/StreamMessage tests). Rebased on latest `main`, 23/23 tests green.

jbonofre commented:
@cshannon @tabish121 I rebased this PR and did again the expected change (that have been changed by a force push 😄 ). If you can do a new pass, that would be great.

cshannon commented:
@jbonofre - Can you do another rebase and bring this up to date again? I should have time soon to get back to this, it would be good to get into 6.4.0

skrcode commented:
JAIPilot Cloud reviewed exact head `5ad7be8834181a68651f270231d71f0e43e2f86a` and found one focused test gap in the changed `ActiveMQObjectMessage` contract.

I opened an AI-assisted, test-only draft against this branch: https://github.com/jbonofre/activemq/pull/1

It verifies that an untrusted serialized payload makes `isBodyAssignableTo()` return `false` and `getBody()` throw `MessageFormatException`, instead of leaking the raw deserialization `JMSException`. The focused JMS 2 suites passed 41 tests and the two affected modules built/compiled cleanly. The draft states the scoped-build limitation and includes explicit model provenance.
